helix-core/src/main/java/org/apache/helix/controller/rebalancer/waged/constraints/BaselineInfluenceConstraint.java

```diff
@@ -0,0 +1,54 @@
+package org.apache.helix.controller.rebalancer.waged.constraints;
+
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import java.util.Map;
+
+import org.apache.helix.controller.rebalancer.waged.model.AssignableNode;
+import org.apache.helix.controller.rebalancer.waged.model.AssignableReplica;
+import org.apache.helix.controller.rebalancer.waged.model.ClusterContext;
+
+
+/**
+ * Evaluate the proposed assignment according to the potential partition movements cost based on
+ * the baseline assignment's influence.
+ * This constraint promotes movements for evenness. If best possible doesn't exist, baseline will be
+ * used to restrict movements, so this constraint should give no score in that case.
+ */
+public class BaselineInfluenceConstraint extends AbstractPartitionMovementConstraint {
+  @Override
+  protected double getAssignmentScore(AssignableNode node, AssignableReplica replica,
+      ClusterContext clusterContext) {
+    Map<String, String> bestPossibleAssignment =
+        getStateMap(replica, clusterContext.getBestPossibleAssignment());
+    Map<String, String> baselineAssignment =
+        getStateMap(replica, clusterContext.getBaselineAssignment());
+    if (bestPossibleAssignment.isEmpty()) {
+      return getMinScore();
+    }
+    return calculateAssignmentScore(node.getInstanceName(), replica.getReplicaState(),
+        baselineAssignment);
+  }
+
+  @Override
+  protected double getStateTransitionCostFactor() {
+    return MAX_SCORE / 2;
```

Comment:
We shall return STATE_TRANSITION_COST_FACTOR here.
1. it should return factor not the actual score.
2. it shall not be hardcoded.

Reply:
Mentioned this in an earlier comment, I will rename the factor to a score, since that was the old behavior. 

I honestly think this way is fine, because it's clearer. The abstract function's methods act as the default implementation: it returns MAX_SCORE if totally match, MAX_SCORE/2 if requires state transtion, MIN_SCORE for anything else. If the extended function wants to be different, it can just override. 

Maybe I can even remove getStateTransitionCostFactor() and just put the default value in calculateAssignmentScore(). What do you think?

Reply:
1. removing getStateTransitionCostFactor() is the same thing as I commented before that we can make this method non-abstract. So no objection on my side.
2. MAX_SCORE / 2 is hardcode. Not good. What if we find that the cost is higher or lower than we thought? Or we want to make it configurable for different clusters? Then we need to fix it anyway.
3. I think the factor design is easier to maintain and cleaner, considering that we are not 100% sure that 0.5 is a good rate for everyone.